File: packages/sdk-snippets/src/targets/python/aws/webhooks/fixtures/security-and-server-variables-create.py
import json

import boto3
from readme_metrics.VerifyWebhook import VerifyWebhook

# Your ReadMe secret as a bytes object; you may want to store this in AWS Secrets Manager
README_SECRET = b"my-readme-secret"

# Your default API Gateway usage plan; this will be attached to the API keys that being created
DEFAULT_USAGE_PLAN_ID = "123abc"


def handler(event, lambda_context):
    status_code = None
    email = None
    api_key = None
    error = None

    try:
        signature = event.get("headers", {}).get("ReadMe-Signature")
        body = json.loads(event.get("body", "{}"))
        VerifyWebhook(body, signature, README_SECRET)

        email = body.get("email")
        client = boto3.client("apigateway")
        keys = client.get_api_keys(nameQuery=email, includeValues=True)
        if len(keys.get("items", [])) > 0:
            # if multiple API keys are returned for the given email, use the first one
            api_key = keys["items"][0]["value"]
            status_code = 200
        else:
            key = client.create_api_key(
                name=email,
                description=f"API key for ReadMe user {email}",
                tags={"user": email, "vendor": "ReadMe"},
                enabled=True,
            )

            client.create_usage_plan_key(
                usagePlanId=DEFAULT_USAGE_PLAN_ID, keyId=key["id"], keyType="API_KEY"
            )

            api_key = key["value"]
            status_code = 200
    except Exception as e:
        error = str(e)
        if error.find("Signature") > -1:
            status_code = 401
        else:
            status_code = 500

    if status_code == 200:
        body = {
            # OAS Server variables
            "name": "default-name",
            "port": "",

            # OAS Security variables
            "petstore_auth": api_key,
            "basic_auth": { "user": email, "pass": api_key },
        }
    else:
        body = {"message": error}

    return {
        "statusCode": status_code,
        "headers": {"Content-Type": "application/json"},
        "body": json.dumps(body),
    }

File: packages/sdk-snippets/src/targets/node/aws/webhooks/fixtures/security-and-server-variables-create.js
const {
  APIGatewayClient,
  CreateApiKeyCommand,
  CreateUsagePlanKeyCommand,
  GetApiKeysCommand,
} = require('@aws-sdk/client-api-gateway');
const readme = require('readmeio');

// Your ReadMe secret; you may want to store this in AWS Secrets Manager
const README_SECRET = 'my-readme-secret';

// Your default API Gateway usage plan; this will be attached to new API keys being created
const DEFAULT_USAGE_PLAN_ID = '123abc';

exports.handler = async event => {
  let statusCode, email, apiKey, error;

  try {
    // Verify the request is legitimate and came from ReadMe.
    const signature = event.headers['ReadMe-Signature'];
    const body = JSON.parse(event.body);
    readme.verifyWebhook(body, signature, README_SECRET);

    // Look up the API key associated with the user's email address.
    const email = body.email;
    const client = new APIGatewayClient();
    const getCommand = new GetApiKeysCommand({ nameQuery: email, includeValues: true });
    const keys = await client.send(getCommand);
    if (keys.items.length > 0) {
      // If multiple API keys are returned for the given email, use the first one.
      apiKey = keys.items[0].value;
      statusCode = 200;
    } else {
      // If no API keys were found, create a new key and apply a usage plan.
      const createKeyCommand = new CreateApiKeyCommand({
        name: email,
        description: `API key for ReadMe user ${email}`,
        tags: {
          user: email,
          vendor: 'ReadMe',
        },
        enabled: true,
      });
      const key = await client.send(createKeyCommand);

      const usagePlanKeyCommand = new CreateUsagePlanKeyCommand({
        usagePlanId: DEFAULT_USAGE_PLAN_ID,
        keyId: key.id,
        keyType: 'API_KEY',
      });
      await client.send(usagePlanKeyCommand);

      apiKey = key.value;
      statusCode = 200;
    }
  } catch (e) {
    error = e.message;
    statusCode = error.match(/Signature/) ? 401 : 500;
  }

  return {
    statusCode,
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      // OAS Server variables
      name: 'default-name',
      port: '',

      // OAS Security variables
      petstore_auth: apiKey,
      basic_auth: { user: email, pass: apiKey },

      // Error message, if any
      message: error,
    }),
  };
};